src/renderer_webgpu.cpp

```diff
@@ -0,0 +1,4808 @@
+	};
+	BX_STATIC_ASSERT(TextureFormat::Count == BX_COUNTOF(s_textureFormat));
+
+	int32_t s_msaa[] =
+	{
+		 1,
+		 2,
+		 4,
+		 8,
+		16,
+	};
+
+	struct RendererContextWgpu;
+	static RendererContextWgpu* s_renderWgpu;
+
+	static bool s_ignoreError = false;
+
+#if !BX_PLATFORM_EMSCRIPTEN
+	DawnSwapChainImplementation(*createSwapChain)(wgpu::Device device, void* nwh);
+
+#ifdef DAWN_ENABLE_BACKEND_D3D12
+	DawnSwapChainImplementation CreateSwapChainD3D12(wgpu::Device device, void* nwh)
+	{
+		HWND win32Window = (HWND)nwh;
+		return dawn_native::d3d12::CreateNativeSwapChainImpl(device.Get(), win32Window);
+	}
+#endif
+
+#ifdef DAWN_ENABLE_BACKEND_VULKAN
+	DawnSwapChainImplementation CreateSwapChainVulkan(wgpu::Device device, void* nwh)
+	{
+		VkInstance instance = dawn_native::vulkan::GetInstance(device.Get());
+
+		PFN_vkCreateWin32SurfaceKHR vkCreateWin32SurfaceKHR = (PFN_vkCreateWin32SurfaceKHR)dawn_native::vulkan::GetInstanceProcAddr(device.Get(), "vkCreateWin32SurfaceKHR");
+
+		VkSurfaceKHR surface;
+#if BX_PLATFORM_WINDOWS
+		// Copied from renderer_vk.cpp -> needs refactor
+		{
+			VkWin32SurfaceCreateInfoKHR sci;
+			sci.sType = VK_STRUCTURE_TYPE_WIN32_SURFACE_CREATE_INFO_KHR;
+			sci.pNext = NULL;
+			sci.flags = 0;
+			sci.hinstance = (HINSTANCE)GetModuleHandle(NULL);
+			sci.hwnd = (HWND)nwh;
+			VkResult result = vkCreateWin32SurfaceKHR(instance, &sci, NULL, &surface);
+		}
+#endif
+		return dawn_native::vulkan::CreateNativeSwapChainImpl(device.Get(), surface);
+	}
+#endif
+
+#endif
+
+	struct RendererContextWgpu : public RendererContextI
+	{
+		RendererContextWgpu()
+			: m_frameIndex(0)
+			, m_numWindows(0)
+			, m_rtMsaa(false)
+			, m_capture(NULL)
+			, m_captureSize(0)
+		{
+			bx::memSet(&m_windows, 0xff, sizeof(m_windows) );
+		}
+
+		~RendererContextWgpu()
+		{
+		}
+
+		bool init(const Init& _init)
+		{
+			BX_UNUSED(_init);
+			BX_TRACE("Init.");
+
+			if (_init.debug
+			||  _init.profile)
+			{
+				m_renderDocDll = loadRenderDoc();
+			}
+
+			setGraphicsDebuggerPresent(NULL != m_renderDocDll);
+
+			m_fbh.idx = kInvalidHandle;
+			bx::memSet(m_uniforms, 0, sizeof(m_uniforms) );
+			bx::memSet(&m_resolution, 0, sizeof(m_resolution) );
+
+#if !BX_PLATFORM_EMSCRIPTEN
+			// Default to D3D12, Metal, Vulkan, OpenGL in that order as D3D12 and Metal are the preferred on
+			// their respective platforms, and Vulkan is preferred to OpenGL
+#if defined(DAWN_ENABLE_BACKEND_D3D12)
+			static dawn_native::BackendType backendType = dawn_native::BackendType::D3D12;
+#elif defined(DAWN_ENABLE_BACKEND_METAL)
+			static dawn_native::BackendType backendType = dawn_native::BackendType::Metal;
+#elif defined(DAWN_ENABLE_BACKEND_OPENGL)
+			static dawn_native::BackendType backendType = dawn_native::BackendType::OpenGL;
+#elif defined(DAWN_ENABLE_BACKEND_VULKAN)
+			static dawn_native::BackendType backendType = dawn_native::BackendType::Vulkan;
+#else
+#error
+#endif
+
+			m_instance.EnableBackendValidation(true);
```

Comment:
This enables things like the Vulkan Validation Layers, you don't want this to be always on.

Reply:
Enabling it only in debug